# CODE JOUIN CLEMENT & YANIS KETO

Attention à certains codes, il faudra redémarrer le noyau pour pouvoir étendre le programme. Pour la partie reconstruction en vidéo, elle est fonctionnelle et constitue un ajout. Il faut cependant attendre environ 1 minute 30 avant de voir la fenêtre apparaître. Nous n'avons pas démarré la partie bonus par manque de temps. Nous avons utilisé le temps supplémentaire pour effectuer la reconstruction de la vidéo.

### 1.1 RÉCUPÉRATION DES FRAMES DE LA VIDÉO ET AFFICHAGE

In [4]:
import cv2
import sys

ESC_KEY = 27
Q_KEY = 113


def poll_key():
    return cv2.waitKey(1) & 0xFF


def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Obtention des propriétés de la vidéo
    fps = cap.get(cv2.CAP_PROP_FPS)
    nb_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    resolution = (int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))

    print(f"FPS de la vidéo: {fps}")
    print(f"Nombre total de frames: {nb_frames}")
    print(f"Résolution de la vidéo: {resolution[0]} x {resolution[1]}")

    # Création des fenêtres OpenCV
    cv2.namedWindow("Vidéo originale")
    cv2.namedWindow("Vidéo en niveaux de gris")

    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir im en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Affichage des images dans les fenêtres
        cv2.imshow("Vidéo originale", im)
        cv2.imshow("Vidéo en niveaux de gris", im_gray)

        # Gestion des événements clavier
        key = poll_key()
        if key == ESC_KEY or key == Q_KEY:
            break

        # Afficher le numéro du frame actuel
        print(f"Frame actuel: {frame_number}")
        frame_number += 1

    # Fermer toutes les fenêtres OpenCV
    cv2.destroyAllWindows()

    return 0


if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
FPS de la vidéo: 30.0
Nombre total de frames: 614
Résolution de la vidéo: 384 x 288
Frame actuel: 0
Frame actuel: 1
Frame actuel: 2
Frame actuel: 3
Frame actuel: 4
Frame actuel: 5
Frame actuel: 6
Frame actuel: 7
Frame actuel: 8
Frame actuel: 9
Frame actuel: 10
Frame actuel: 11
Frame actuel: 12
Frame actuel: 13
Frame actuel: 14
Frame actuel: 15
Frame actuel: 16
Frame actuel: 17
Frame actuel: 18
Frame actuel: 19
Frame actuel: 20
Frame actuel: 21
Frame actuel: 22
Frame actuel: 23
Frame actuel: 24
Frame actuel: 25
Frame actuel: 26
Frame actuel: 27
Frame actuel: 28
Frame actuel: 29
Frame actuel: 30
Frame actuel: 31
Frame actuel: 32
Frame actuel: 33
Frame actuel: 34
Frame actuel: 35
Frame actuel: 36
Frame actuel: 37
Frame actuel: 38
Frame actuel: 39
Frame actuel: 40
Frame actuel: 41
Frame actuel: 42
Frame actuel: 43
Frame actuel: 44
Frame actuel: 45
Frame actuel: 46
Frame actuel: 47
Frame actuel: 48
Frame actuel: 49
Frame actuel: 5

SystemExit: 0

![Image Description](./captures/part1.png)


##### Questions
Question 2 : Ils y a 614 frames; \
Question 3 : A raison de 30 fps, la vidéo dure 20,4s; \
Question 4 : Résolution de la vidéo: 384 x 288; \
Question 7 : OpenCV stocke les images couleur en utilisant le modèle de couleur BGR (bleu, vert, rouge), contrairement à d'autres modèles de couleur couramment utilisés tels que RGB (rouge, vert, bleu). Cela signifie que, pour chaque pixel de l'image, les intensités des composantes bleue (B), verte (G), et rouge (R) sont stockées dans cet ordre.

#### 1.3 ENREGISTREMENT DES FRAMES DE LA VIDÉO DANS UN VECTEUR

In [5]:
import cv2
import sys

ESC_KEY = 27
Q_KEY = 113


def poll_key():
    return cv2.waitKey(1) & 0xFF


def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Obtention des propriétés de la vidéo
    fps = cap.get(cv2.CAP_PROP_FPS)
    nb_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    resolution = (int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))

    print(f"FPS de la vidéo: {fps}")
    print(f"Nombre total de frames: {nb_frames}")
    print(f"Résolution de la vidéo: {resolution[0]} x {resolution[1]}")

    # Création d'une liste pour stocker les images en niveaux de gris
    images_vector = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir im en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        images_vector.append(im_gray)

        # Affichage des images dans les fenêtres (commenté pour éviter l'affichage en temps réel)
        # cv2.imshow("Vidéo originale", im)
        # cv2.imshow("Vidéo en niveaux de gris", im_gray)

        # Gestion des événements clavier
        key = poll_key()
        if key == ESC_KEY or key == Q_KEY:
            break

        # Afficher le numéro du frame actuel
        print(f"Frame actuel: {frame_number}")
        frame_number += 1

    # Fermer la fenêtre OpenCV (commenté car non utilisé)
    # cv2.destroyAllWindows()

    # Afficher le nombre d'images stockées dans le vecteur
    print(f"Nombre d'images stockées dans le vecteur: {len(images_vector)}")

    # Afficher les images contenues dans le vecteur
    for i, image in enumerate(images_vector):
        cv2.imshow(f"Image {i + 1}", image)
        key = cv2.waitKey(0) & 0xFF
        if key == ESC_KEY or key == Q_KEY:
            break

    cv2.destroyAllWindows()

    return 0


if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
FPS de la vidéo: 30.0
Nombre total de frames: 614
Résolution de la vidéo: 384 x 288
Frame actuel: 0
Frame actuel: 1
Frame actuel: 2
Frame actuel: 3
Frame actuel: 4
Frame actuel: 5
Frame actuel: 6
Frame actuel: 7
Frame actuel: 8
Frame actuel: 9
Frame actuel: 10
Frame actuel: 11
Frame actuel: 12
Frame actuel: 13
Frame actuel: 14
Frame actuel: 15
Frame actuel: 16
Frame actuel: 17
Frame actuel: 18
Frame actuel: 19
Frame actuel: 20
Frame actuel: 21
Frame actuel: 22
Frame actuel: 23
Frame actuel: 24
Frame actuel: 25
Frame actuel: 26
Frame actuel: 27
Frame actuel: 28
Frame actuel: 29
Frame actuel: 30
Frame actuel: 31
Frame actuel: 32
Frame actuel: 33
Frame actuel: 34
Frame actuel: 35
Frame actuel: 36
Frame actuel: 37
Frame actuel: 38
Frame actuel: 39
Frame actuel: 40
Frame actuel: 41
Frame actuel: 42
Frame actuel: 43
Frame actuel: 44
Frame actuel: 45
Frame actuel: 46
Frame actuel: 47
Frame actuel: 48
Frame actuel: 49
Frame actuel: 5

SystemExit: 0

![Image Description](./captures/part1.2.png)

### 1.4 EXTRACTION DU FOND DE LA VIDÉO ET DE LA ROUTE


In [22]:
import cv2
import sys

ESC_KEY = 27
Q_KEY = 113
M = 5  # Nombre d'images pour calculer la moyenne

def poll_key():
    return cv2.waitKey(1) & 0xFF


def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Obtention des propriétés de la vidéo
    fps = cap.get(cv2.CAP_PROP_FPS)
    nb_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    resolution = (int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))

    print(f"FPS de la vidéo: {fps}")
    print(f"Nombre total de frames: {nb_frames}")
    print(f"Résolution de la vidéo: {resolution[0]} x {resolution[1]}")

    # Création d'une liste pour stocker les images en niveaux de gris
    images_vector = []
    
    # Création d'une liste pour stocker la moyenne des images sur M frames
    average_images = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir im en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        images_vector.append(im_gray)

        # Si le nombre d'images atteint M, calculer la moyenne
        if len(images_vector) == M:
            # Effectuer la somme des images avec un type de données plus large
            sum_images = sum(images_vector, 0.0)
            
            # Calculer la moyenne en normalisant la somme
            average_image = (sum_images / M).astype('uint8')
            
            # Ajouter l'image moyenne à la liste
            average_images.append(average_image)
            
            # Réinitialiser la liste pour les prochaines M images
            images_vector = []

        # Gestion des événements clavier
        key = poll_key()
        if key == ESC_KEY or key == Q_KEY:
            break

        # Afficher le numéro du frame actuel
        print(f"Frame actuel: {frame_number}")
        frame_number += 1

    # Fermer la fenêtre OpenCV (commenté car non utilisé)
    # cv2.destroyAllWindows()

    # Afficher le nombre d'images stockées dans le vecteur
    print(f"Nombre d'images stockées dans le vecteur: {len(average_images)}")

    # Afficher les images moyennes
    for i, average_image in enumerate(average_images):
        cv2.imshow(f"Image Moyenne {i + 1}", average_image)
        key = cv2.waitKey(0) & 0xFF
        if key == ESC_KEY or key == Q_KEY:
            break

    cv2.destroyAllWindows()

    return 0


if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
FPS de la vidéo: 30.0
Nombre total de frames: 614
Résolution de la vidéo: 384 x 288
Frame actuel: 0
Frame actuel: 1
Frame actuel: 2
Frame actuel: 3
Frame actuel: 4
Frame actuel: 5
Frame actuel: 6
Frame actuel: 7
Frame actuel: 8
Frame actuel: 9
Frame actuel: 10
Frame actuel: 11
Frame actuel: 12
Frame actuel: 13
Frame actuel: 14
Frame actuel: 15
Frame actuel: 16
Frame actuel: 17
Frame actuel: 18
Frame actuel: 19
Frame actuel: 20
Frame actuel: 21
Frame actuel: 22
Frame actuel: 23
Frame actuel: 24
Frame actuel: 25
Frame actuel: 26
Frame actuel: 27
Frame actuel: 28
Frame actuel: 29
Frame actuel: 30
Frame actuel: 31
Frame actuel: 32
Frame actuel: 33
Frame actuel: 34
Frame actuel: 35
Frame actuel: 36
Frame actuel: 37
Frame actuel: 38
Frame actuel: 39
Frame actuel: 40
Frame actuel: 41
Frame actuel: 42
Frame actuel: 43
Frame actuel: 44
Frame actuel: 45
Frame actuel: 46
Frame actuel: 47
Frame actuel: 48
Frame actuel: 49
Frame actuel: 5

SystemExit: 0

![image.png](./captures/part2.1-200.png)

Question 9 : Il n'est pas simple de determiner la bonne valeur de M pour le moyennage de nos images, cela dépend de la vidéo et du contexte de celle-ci. 

### 1.4.3 EXTRACTION DE LA ROUTE DE LA SÉQUENCE VIDÉO

Pour cela nous avons essayé beaucoup de choses retrouvé en commentaire, le filtre de canny n'est pas pertinent. Noous avons dans un premier temps fait un filtre rapide pour nous permettre d'avancer sur le reste du sujet mais il n'était pas assez restrictif et nous à poser problème pour la suite. Nous sommes donc revenus avec un autre filtre plus appliqué qui nous permets de mieux isoler la route.

La partie code test avec nouveau filtre présente plus en détails la raison d'utilisation des filtres.

In [ ]:
import cv2
import sys
import numpy as np

ESC_KEY = 27
Q_KEY = 113
M = 200  # Nombre d'images pour calculer la moyenne

def poll_key():
    return cv2.waitKey(1) & 0xFF

# def detect_road(frame):
#     # Appliquer un filtre Canny pour détecter les contours
#     edges = cv2.Canny(frame, 50, 100)  # Seuil inférieur (low_threshold) : 0, Seuil supérieur (high_threshold) : 250

#     # Obtenez les dimensions de l'image
#     height, width = frame.shape[:2]

#     # Créer une image noire de la même taille que les contours détectés
#     mask = np.zeros_like(edges)

#     # Définir les points du polygone qui représente la région d'intérêt (ROI) - ici, un trapèze
#     roi_vertices = np.array([[(0, height), (0, 0), (width, 0), (width, height)]], dtype=np.int32)

#     # Remplir le polygone avec des pixels blancs (255) pour créer la masque
#     cv2.fillPoly(mask, roi_vertices, 255)

#     # Appliquer l'opération AND entre les contours et la masque pour isoler la région d'intérêt
#     masked_edges = cv2.bitwise_and(edges, mask)

#     return masked_edges

def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Obtention des propriétés de la vidéo
    fps = cap.get(cv2.CAP_PROP_FPS)
    nb_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    resolution = (int(cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))

    print(f"FPS de la vidéo: {fps}")
    print(f"Nombre total de frames: {nb_frames}")
    print(f"Résolution de la vidéo: {resolution[0]} x {resolution[1]}")

    # Création d'une liste pour stocker les images en niveaux de gris
    images_vector = []
    
    # Création d'une liste pour stocker la moyenne des images sur M frames
    average_images = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir im en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        images_vector.append(im_gray)

        # Si le nombre d'images atteint M, calculer la moyenne
        if len(images_vector) == M:
            # Effectuer la somme des images avec un type de données plus large
            sum_images = sum(images_vector, 0.0)
            
            # Calculer la moyenne en normalisant la somme
            average_image = (sum_images / M).astype('uint8')
        
            # Ajouter l'image moyenne à la liste
            average_images.append(average_image)
            
            # Réinitialiser la liste pour les prochaines M images
            images_vector = []

            key = cv2.waitKey(0) & 0xFF
            if key == ESC_KEY or key == Q_KEY:
                break

        # Gestion des événements clavier
        key = poll_key()
        if key == ESC_KEY or key == Q_KEY:
            break

        # Afficher le numéro du frame actuel
        print(f"Frame actuel: {frame_number}")
        frame_number += 1

    # Fermer la fenêtre OpenCV (commenté car non utilisé)
    # cv2.destroyAllWindows()

    # Afficher le nombre d'images stockées dans le vecteur
    print(f"Nombre d'images stockées dans le vecteur: {len(average_images)}")

    # Afficher les images moyennes avec la détection de la route
    for i, average_image in enumerate(average_images):
        cv2.imshow(f"Image Moyenne {i + 1}", average_image)

        # Appliquer le seuillage
        _, thresholded_image_low = cv2.threshold(average_image, 90, 255, cv2.THRESH_BINARY)
        # cv2.imshow("Thresholded Image low", thresholded_image_low)


        _, thresholded_image_high = cv2.threshold(average_image, 195, 255, cv2.THRESH_BINARY)
        # cv2.imshow("Thresholded Image high", thresholded_image_high)

        # road_detected = detect_road(thresholded_image)
        # cv2.imshow(f"Image Moyenne avec Détection de Route", road_detected)

        subtracted_image = cv2.absdiff(thresholded_image_low, thresholded_image_high)
        # cv2.imshow("Subtracted Image", subtracted_image)

        # Appliquer la dilation
        dilated_image = cv2.dilate(subtracted_image, None, iterations=1)
        # cv2.imshow("Dilated Image", dilated_image)

        # Appliquer l'érosion
        eroded_image = cv2.erode(dilated_image, None, iterations=9)
        cv2.imshow("Eroded Image", eroded_image)

        key = cv2.waitKey(0) & 0xFF
        if key == ESC_KEY or key == Q_KEY:
            break

    cv2.destroyAllWindows()

    return 0

if __name__ == "__main__":
    sys.exit(main())


![image.png](./captures/filtre.png)

#### Isololer les voitures à l'aide de filtre

In [1]:
import cv2
import sys
import numpy as np

ESC_KEY = 27
Q_KEY = 113
M = 200  # Nombre d'images pour calculer la moyenne

def poll_key():
    return cv2.waitKey(1) & 0xFF

def isolate_cars(img_vector_car, road_mask):
    isolated_cars_vector = []
    for idx, img in enumerate(img_vector_car):
        _, thresholded_frame = cv2.threshold(img, 80, 255, cv2.THRESH_BINARY)
        isolated_cars = cv2.bitwise_and(thresholded_frame, road_mask)
        isolated_cars_vector.append(isolated_cars)
    return isolated_cars_vector

def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Création d'une liste pour stocker les images en niveaux de gris
    images_vector = []

    img_vector_car = []

    # Création d'une liste pour stocker la moyenne des images sur M frames
    average_images = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir l'image en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        images_vector.append(im_gray)
        img_vector_car.append(im_gray)

        # Si le nombre total d'images atteint M, calculer la moyenne
        if frame_number >= M:
            # Effectuer la somme des images avec un type de données plus large
            sum_images = sum(images_vector, 0.0)

            # Calculer la moyenne en normalisant la somme
            average_image = (sum_images / M).astype('uint8')

            # Ajouter l'image moyenne à la liste
            average_images.append(average_image)

            # Réinitialiser la liste pour les prochaines M images
            images_vector = []

            key = cv2.waitKey(0) & 0xFF
            if key == ESC_KEY or key == Q_KEY:
                break

            # Afficher le numéro du frame actuel
            print(f"Frame actuel: {frame_number}")

            _, thresholded_image_low = cv2.threshold(average_image, 90, 255, cv2.THRESH_BINARY)
            cv2.imshow("Thresholded Image low", thresholded_image_low)

            _, thresholded_image_high = cv2.threshold(average_image, 195, 255, cv2.THRESH_BINARY)
            cv2.imshow("Thresholded Image high", thresholded_image_high)

            subtracted_image = cv2.absdiff(thresholded_image_low, thresholded_image_high)
            cv2.imshow("Subtracted Image", subtracted_image)

            dilated_image = cv2.dilate(subtracted_image, None, iterations=1)
            cv2.imshow("Dilated Image", dilated_image)

            eroded_image = cv2.erode(dilated_image, None, iterations=9)
            cv2.imshow("Eroded Image", eroded_image)
            road_mask = eroded_image

            # Isoler les voitures dans l'image courante en utilisant le masque de route
            isolated_cars_vector = isolate_cars(img_vector_car, road_mask)

            # Afficher chaque image isolée
            cv2.imshow(f"Isolated Cars - Frame {1}", isolated_cars_vector[0])

        frame_number += 1

    cv2.destroyAllWindows()
    return 0

if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
Frame actuel: 200


KeyboardInterrupt: 

![image.png](./captures/part3.1-200.png)

#### Reconstruire une vidéo 
nous avons reconstruit une vidéo à partir des images cela peut-etre un peu long avant d'avoir la fenêtre environs 1m30

In [1]:
import cv2
import sys
import numpy as np

ESC_KEY = 27
Q_KEY = 113
M = 200  # Nombre d'images pour calculer la moyenne

def isolate_cars(img_vector_car, road_mask):
    isolated_cars_vector = []
    for img in img_vector_car:
        _, thresholded_frame = cv2.threshold(img, 80, 255, cv2.THRESH_BINARY)
        isolated_cars = cv2.bitwise_and(thresholded_frame, road_mask)
        isolated_cars_vector.append(isolated_cars)
    return isolated_cars_vector

def count_cars(isolated_image):
    # Trouver les contours dans l'image isolée
    contours, _ = cv2.findContours(isolated_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Initialiser le compteur de voitures
    car_count = 0

    # Parcourir tous les contours et compter les voitures
    for contour in contours:
        area = cv2.contourArea(contour)
        if area > 50:  # Filtre pour exclure les petits contours parasites
            car_count += 1

    return car_count

def save_images_as_video(images, video_name, fps=30):
    if not images:
        # Si la liste est vide, utilisez une taille par défaut
        height, width = 480, 640
    else:
        height, width = images[0].shape

    fourcc = cv2.VideoWriter_fourcc(*'MJPG')
    out = cv2.VideoWriter(video_name, fourcc, fps, (width, height), isColor=False)

    for image in images:
        out.write(image)

    out.release()

def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Création d'une liste pour stocker les images en niveaux de gris
    img_vector_car = []

    # Création d'une liste pour stocker les images isolées
    isolated_images_vector = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir l'image en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        img_vector_car.append(im_gray)

        # Si le nombre total d'images atteint M, isoler les voitures
        if len(img_vector_car) >= M:
            avg_image = cv2.reduce(np.array(img_vector_car[-M:], dtype=np.uint8), 0, cv2.REDUCE_AVG)[0]
            
            _, thresholded_image_low = cv2.threshold(avg_image, 90, 255, cv2.THRESH_BINARY)
            _, thresholded_image_high = cv2.threshold(avg_image, 195, 255, cv2.THRESH_BINARY)
            
            subtracted_image = cv2.absdiff(thresholded_image_low, thresholded_image_high)
            road_mask = cv2.erode(cv2.dilate(subtracted_image, None, iterations=1), None, iterations=9)

            # Isoler les voitures dans l'image courante en utilisant le masque de route
            isolated_cars_vector = isolate_cars(img_vector_car[-M:], road_mask)

            # Compter et afficher le nombre de voitures
            for idx, isolated_cars in enumerate(isolated_cars_vector):
                car_count = count_cars(isolated_cars)
                print(f"Frame {frame_number - M + idx} - Nombre de voitures: {car_count}")

            # Ajouter les images isolées à la liste
            isolated_images_vector.extend(isolated_cars_vector)

        key = cv2.waitKey(1) & 0xFF
        if key == ESC_KEY or key == Q_KEY:
            break

        frame_number += 1

    cap.release()
    cv2.destroyAllWindows()

    # Enregistrer les images isolées en tant que vidéo
    save_images_as_video(isolated_images_vector, 'output_video.avi')

    # Ouvrir la vidéo générée
    generated_video = cv2.VideoCapture('output_video.avi')
    while True:
        ret, frame = generated_video.read()
        if not ret:
            break

        cv2.imshow('Generated Video', frame)
        if cv2.waitKey(30) & 0xFF == 27:
            break

    generated_video.release()
    cv2.destroyAllWindows()

    return 0

if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
Frame -1 - Nombre de voitures: 4
Frame 0 - Nombre de voitures: 3
Frame 1 - Nombre de voitures: 5
Frame 2 - Nombre de voitures: 5
Frame 3 - Nombre de voitures: 4
Frame 4 - Nombre de voitures: 4
Frame 5 - Nombre de voitures: 4
Frame 6 - Nombre de voitures: 4
Frame 7 - Nombre de voitures: 3
Frame 8 - Nombre de voitures: 5
Frame 9 - Nombre de voitures: 5
Frame 10 - Nombre de voitures: 4
Frame 11 - Nombre de voitures: 5
Frame 12 - Nombre de voitures: 4
Frame 13 - Nombre de voitures: 4
Frame 14 - Nombre de voitures: 4
Frame 15 - Nombre de voitures: 4
Frame 16 - Nombre de voitures: 4
Frame 17 - Nombre de voitures: 4
Frame 18 - Nombre de voitures: 4
Frame 19 - Nombre de voitures: 4
Frame 20 - Nombre de voitures: 4
Frame 21 - Nombre de voitures: 4
Frame 22 - Nombre de voitures: 6
Frame 23 - Nombre de voitures: 6
Frame 24 - Nombre de voitures: 8
Frame 25 - Nombre de voitures: 6
Frame 26 - Nombre de voitures: 7
Frame 27 - Nombre de voit


KeyboardInterrupt

C:\Users\cleme\AppData\Roaming\Python\Python311\site-packages\IPython\core\interactiveshell.py:3516: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


![image.png](./captures/part3-200.png)

#### Comptage de voiture

In [2]:
import cv2
import sys
import numpy as np

ESC_KEY = 27
Q_KEY = 113
M = 200  # Nombre d'images pour calculer la moyenne

def isolate_cars(img_vector_car, road_mask):
    isolated_cars_vector = []
    for img in img_vector_car:
        _, thresholded_frame = cv2.threshold(img, 80, 255, cv2.THRESH_BINARY)
        isolated_cars = cv2.bitwise_and(thresholded_frame, road_mask)
        isolated_cars_vector.append(isolated_cars)
    return isolated_cars_vector

def count_cars(isolated_image):
    # Trouver les contours dans l'image isolée
    contours, _ = cv2.findContours(isolated_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Initialiser le compteur de voitures
    car_count = 0

    # Parcourir tous les contours et compter les voitures
    for contour in contours:
        area = cv2.contourArea(contour)
        if area > 50:  # Filtre pour exclure les petits contours parasites
            car_count += 1

    return car_count

def save_images_as_video(images, video_name, fps=30):
    if not images:
        # Si la liste est vide, utilisez une taille par défaut
        height, width = 480, 640
    else:
        height, width = images[0].shape

    fourcc = cv2.VideoWriter_fourcc(*'MJPG')
    out = cv2.VideoWriter(video_name, fourcc, fps, (width, height), isColor=False)

    for image in images:
        out.write(image)

    out.release()

def main():
    # Chemin vers la vidéo
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    # Ouverture de la capture vidéo
    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    # Création d'une liste pour stocker les images en niveaux de gris
    img_vector_car = []

    # Création d'une liste pour stocker les images isolées
    isolated_images_vector = []

    # Boucle principale pour lire chaque frame de la vidéo
    frame_number = 0
    while True:
        # Lecture du frame suivant
        ret, im = cap.read()

        if not ret:
            # Impossible de lire le frame suivant, sortie de la boucle
            break

        # Convertir l'image en niveaux de gris
        im_gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)

        # Ajouter l'image en niveaux de gris à la liste
        img_vector_car.append(im_gray)

        # Si le nombre total d'images atteint M, isoler les voitures
        if len(img_vector_car) >= M:
            avg_image = cv2.reduce(np.array(img_vector_car[-M:], dtype=np.uint8), 0, cv2.REDUCE_AVG)[0]
            
            _, thresholded_image_low = cv2.threshold(avg_image, 90, 255, cv2.THRESH_BINARY)
            _, thresholded_image_high = cv2.threshold(avg_image, 195, 255, cv2.THRESH_BINARY)
            
            subtracted_image = cv2.absdiff(thresholded_image_low, thresholded_image_high)
            road_mask = cv2.erode(cv2.dilate(subtracted_image, None, iterations=1), None, iterations=9)

            # Isoler les voitures dans l'image courante en utilisant le masque de route
            isolated_cars_vector = isolate_cars(img_vector_car[-M:], road_mask)

            # Compter et afficher le nombre de voitures
            for idx, isolated_cars in enumerate(isolated_cars_vector):
                car_count = count_cars(isolated_cars)
                print(f"Frame {frame_number - M + idx} - Nombre de voitures: {car_count}")

            # Ajouter les images isolées à la liste
            isolated_images_vector.extend(isolated_cars_vector)

            # Sélectionner une image isolée pour affichage
            selected_frame_index = 0  # Choisissez l'index de la frame que vous souhaitez visualiser
            selected_frame = isolated_images_vector[selected_frame_index]

            # Trouver les contours dans l'image isolée
            contours, _ = cv2.findContours(selected_frame, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

            # Initialiser le compteur de voitures
            car_count_selected = 0

            # Parcourir tous les contours et compter les voitures
            for contour in contours:
                area = cv2.contourArea(contour)
                if area > 50:  # Filtre pour exclure les petits contours parasites
                    car_count_selected += 1

            # Dessiner les contours sur l'image pour visualisation
            contours_image = cv2.cvtColor(selected_frame, cv2.COLOR_GRAY2BGR)
            cv2.drawContours(contours_image, contours, -1, (0, 255, 0), 2)

            # Afficher le nombre de voitures
            cv2.putText(contours_image, f"Nombre de voitures: {car_count_selected}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)

            # Afficher l'image résultante
            cv2.imshow("Résultat avec le nombre de voitures", contours_image)
            cv2.waitKey(0)
            cv2.destroyAllWindows()

        key = cv2.waitKey(1) & 0xFF
        if key == ESC_KEY or key == Q_KEY:
            break

        frame_number += 1

    cap.release()
    cv2.destroyAllWindows()

    # Enregistrer les images isolées en tant que vidéo
    save_images_as_video(isolated_images_vector, 'output_video.avi')

    # Ouvrir la vidéo générée
    generated_video = cv2.VideoCapture('output_video.avi')
    while True:
        ret, frame = generated_video.read()
        if not ret:
            break

        cv2.imshow('Generated Video', frame)
        if cv2.waitKey(30) & 0xFF == 27:
            break

    generated_video.release()
    cv2.destroyAllWindows()

    return 0

if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
Frame -1 - Nombre de voitures: 4
Frame 0 - Nombre de voitures: 3
Frame 1 - Nombre de voitures: 5
Frame 2 - Nombre de voitures: 5
Frame 3 - Nombre de voitures: 4
Frame 4 - Nombre de voitures: 4
Frame 5 - Nombre de voitures: 4
Frame 6 - Nombre de voitures: 4
Frame 7 - Nombre de voitures: 3
Frame 8 - Nombre de voitures: 5
Frame 9 - Nombre de voitures: 5
Frame 10 - Nombre de voitures: 4
Frame 11 - Nombre de voitures: 5
Frame 12 - Nombre de voitures: 4
Frame 13 - Nombre de voitures: 4
Frame 14 - Nombre de voitures: 4
Frame 15 - Nombre de voitures: 4
Frame 16 - Nombre de voitures: 4
Frame 17 - Nombre de voitures: 4
Frame 18 - Nombre de voitures: 4
Frame 19 - Nombre de voitures: 4
Frame 20 - Nombre de voitures: 4
Frame 21 - Nombre de voitures: 4
Frame 22 - Nombre de voitures: 6
Frame 23 - Nombre de voitures: 6
Frame 24 - Nombre de voitures: 8
Frame 25 - Nombre de voitures: 6
Frame 26 - Nombre de voitures: 7
Frame 27 - Nombre de voit

![image.png](./captures/par4-200.png)

### Code test avec nouveau filtre 
Pour finir, nous avons essayé de modifier les filtres. Cependant, du fait que nous avons modifié le filtre pour isoler la route, le filtre pour détecter les voitures sur la route ne fonctionne plus. En ce qui concerne les filtres, nous les avons moyennés avec la fonction de seuillage afin d'identifier la route avec un seuillage adapté. Nous avons appliqué la même approche au ciel et à une partie des bâtiments, puis nous avons soustrait les deux images. De plus, nous avons utilisé des filtres bilatéraux et gaussiens pour lisser la zone, notamment celle de la route, suivi d'une opération d'érosion. Cette méthode est similaire à celle utilisée au début du TP, à l'exception des filtres lisseurs.

In [2]:
import cv2
import sys
import numpy as np

ESC_KEY = 27
Q_KEY = 113

def isolate_cars(img_vector_car, road_mask):
    isolated_cars_vector = []
    for img in img_vector_car:
        _, thresholded_frame = cv2.threshold(img, 80, 255, cv2.THRESH_BINARY)
        isolated_cars = cv2.bitwise_and(thresholded_frame, road_mask)
        isolated_cars_vector.append(isolated_cars)
    return isolated_cars_vector

def yanisbestfilter(img_background):
    im = cv2.equalizeHist(img_background)
    sharpening_kernel = np.array([[-1, -1, -1],
                                [-1, 9, -1],
                                [-1, -1, -1]])

   # Appliquer le noyau à l'image
    im = cv2.GaussianBlur(im, (5, 5), 0)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.GaussianBlur(im, (5, 5), 0)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.GaussianBlur(im, (5, 5), 0)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.GaussianBlur(im, (5, 5), 0)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.GaussianBlur(im, (5, 5), 0)
    im = cv2.bilateralFilter(im, 31, 25, 75)
    im = cv2.bilateralFilter(im, 31, 25, 75)

    im = cv2.filter2D(im, -1, sharpening_kernel)

    _, road_sky = cv2.threshold(im, 130, 255, cv2.THRESH_BINARY)
    _, sky = cv2.threshold(im, 210, 255, cv2.THRESH_BINARY)

    sky = cv2.erode(sky, None, iterations=1)
    output = cv2.bitwise_xor(road_sky, sky)

    output = cv2.erode(output, None, iterations=4)
    output = cv2.GaussianBlur(output, (5, 5), 0)
    _, output = cv2.threshold(output, 220, 255, cv2.THRESH_BINARY)

    output = cv2.dilate(output, None, iterations=8)

    kernel_size = 31
    kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (kernel_size, kernel_size))

    output = cv2.dilate(output, kernel, iterations=2)
    output = cv2.erode(output, kernel, iterations=2)

    return output

def count_cars(isolated_image):
    contours, _ = cv2.findContours(isolated_image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    car_count = 0

    for contour in contours:
        area = cv2.contourArea(contour)
        if area > 50:
            car_count += 1

    return car_count

def extract_background(video_flow, frame_average):
    if len(video_flow) < frame_average:
        print("Le flux vidéo ne contient pas assez d'images.")
        return None
    
    avg_background = np.zeros(video_flow[0].shape, np.float32)

    for i in range(1, frame_average):
        cv2.accumulateWeighted(np.float32(video_flow[i]), avg_background, 1.0 / frame_average)
    background = cv2.convertScaleAbs(avg_background)

    return background

def save_images_as_video(images, video_name, fps=30):
    if not images:
        height, width = 480, 640
    else:
        height, width = images[0].shape

    fourcc = cv2.VideoWriter_fourcc(*'MJPG')
    out = cv2.VideoWriter(video_name, fourcc, fps, (width, height), isColor=False)

    for image in images:
        out.write(image)

    out.release()

def main():
    video_name = "./Fichiers_TP4/Fichiers_autoroute/video.avi"
    print("Nom de la vidéo:", video_name)

    cap = cv2.VideoCapture(video_name)
    if not cap.isOpened():
        print("L'ouverture de la capture a échoué; nous ne pouvons rien faire :'(")
        return 1

    grey_img_vector = []
    isolated_cars_vector = []

    frame_number = 0

    while True:
        ret, capture_frame = cap.read()

        if not ret:
            break

        grey_frame = cv2.cvtColor(capture_frame, cv2.COLOR_BGR2GRAY)
        grey_img_vector.append(grey_frame)

    avg_image = extract_background(grey_img_vector, 600)
    if avg_image is None:
        print("Impossible de calculer l'image moyenne.")
        return 1

    road_mask = yanisbestfilter(avg_image)

    cv2.imshow("Average image ", avg_image)
    cv2.imshow("road mask", road_mask)

    isolated_cars_vector = isolate_cars(grey_img_vector, road_mask)

    for idx, isolated_cars in enumerate(isolated_cars_vector):
        car_count = count_cars(isolated_cars)
        print(f"Frame {frame_number + idx} - Nombre de voitures: {car_count}")

    selected_frame_index = 0
    selected_frame = isolated_cars_vector[selected_frame_index]
    contours, _ = cv2.findContours(selected_frame, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    
    car_count_selected = 0
    for contour in contours:
        area = cv2.contourArea(contour)
        if area > 50:
            car_count_selected += 1

    contours_image = cv2.cvtColor(selected_frame, cv2.COLOR_GRAY2BGR)
    cv2.drawContours(contours_image, contours, -1, (0, 255, 0), 2)

    cv2.imshow("Resultat avec le nombre de voitures", contours_image)

    key = cv2.waitKey(1) & 0xFF
    if key == ESC_KEY or key == Q_KEY:
        cap.release()
        cv2.destroyAllWindows()
        save_images_as_video(isolated_cars_vector, 'output_video.avi')
        generated_video = cv2.VideoCapture('output_video.avi')
        
        while True:
            ret, frame = generated_video.read()
            if not ret:
                break

            cv2.imshow('Generated Video', frame)
            if cv2.waitKey(30) & 0xFF == 27:
                break

        generated_video.release()
        cv2.destroyAllWindows()
        return 0

if __name__ == "__main__":
    sys.exit(main())


Nom de la vidéo: ./Fichiers_TP4/Fichiers_autoroute/video.avi
Frame 0 - Nombre de voitures: 2
Frame 1 - Nombre de voitures: 2
Frame 2 - Nombre de voitures: 4
Frame 3 - Nombre de voitures: 3
Frame 4 - Nombre de voitures: 3
Frame 5 - Nombre de voitures: 2
Frame 6 - Nombre de voitures: 2
Frame 7 - Nombre de voitures: 2
Frame 8 - Nombre de voitures: 2
Frame 9 - Nombre de voitures: 2
Frame 10 - Nombre de voitures: 2
Frame 11 - Nombre de voitures: 2
Frame 12 - Nombre de voitures: 2
Frame 13 - Nombre de voitures: 2
Frame 14 - Nombre de voitures: 2
Frame 15 - Nombre de voitures: 2
Frame 16 - Nombre de voitures: 2
Frame 17 - Nombre de voitures: 2
Frame 18 - Nombre de voitures: 2
Frame 19 - Nombre de voitures: 2
Frame 20 - Nombre de voitures: 2
Frame 21 - Nombre de voitures: 2
Frame 22 - Nombre de voitures: 2
Frame 23 - Nombre de voitures: 4
Frame 24 - Nombre de voitures: 4
Frame 25 - Nombre de voitures: 7
Frame 26 - Nombre de voitures: 5
Frame 27 - Nombre de voitures: 4
Frame 28 - Nombre de voit

SystemExit: 

: 